File: app/src/test/java/com/kira/kdownloader/settings/FilenameTemplateTest.java
package com.kira.kdownloader.settings;

import org.junit.Test;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

public class FilenameTemplateTest {
    private final FilenameTemplate.Values values = new FilenameTemplate.Values(
            "Song Title", "Artist", "2026-07-20", "1080p", "mp4", "abc123");

    @Test public void substitutesAllVariables() {
        assertEquals("Artist - Song Title [1080p]", FilenameTemplate.render("{channel} - {title} [{quality}]", values, 250));
    }

    @Test public void removesCharactersAndroidRejects() {
        String result = FilenameTemplate.render("{title}", values.withTitle("a/b:c*?\"<>|d"), 250);
        for (char c : "\\/:*?\"<>|".toCharArray()) assertTrue(result.indexOf(c) < 0);
    }

    @Test public void validateRejectsUnknownVariable() {
        assertTrue(FilenameTemplate.validate("{title}-{bogus}") instanceof FilenameTemplate.Validation.Invalid);
    }

    @Test public void validateRejectsUnbalancedBraces() {
        assertTrue(FilenameTemplate.validate("{title") instanceof FilenameTemplate.Validation.Invalid);
        assertTrue(FilenameTemplate.validate("title}") instanceof FilenameTemplate.Validation.Invalid);
    }

    @Test public void validateAcceptsAWellFormedTemplate() {
        assertEquals(FilenameTemplate.Validation.Valid, FilenameTemplate.validate("{title} {id}"));
    }

    @Test public void renderFallsBackToTitleForInvalidTemplate() {
        assertEquals("Song Title", FilenameTemplate.render("{unknown}", values, 250));
    }

    @Test public void clampPreservesATrailingExtension() {
        StringBuilder name = new StringBuilder();
        for (int i = 0; i < 300; i++) name.append('x');
        String result = FilenameTemplate.render(name + ".mp4", values, 20);
        assertTrue(result.endsWith(".mp4"));
        assertTrue(result.length() <= 20);
    }

    @Test public void renderNeverReturnsBlank() {
        assertTrue(!FilenameTemplate.render("{title}", values.withTitle("***"), 120).trim().isEmpty());
    }

    @Test public void sanitizeCollapsesRepeatedUnderscoresAndTrims() {
        assertEquals("a_b", FilenameTemplate.sanitize("__a///b__"));
    }

    @Test public void exampleIsNotBlank() {
        assertTrue(!FilenameTemplate.example("{title}-{channel}", 120).trim().isEmpty());
    }
}

File: app/src/test/java/com/kira/kdownloader/settings/FilenameTemplateTest.kt
package com.kira.kdownloader.settings

import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

class FilenameTemplateTest {

    private val values = FilenameTemplate.Values(
        title = "Song Title",
        channel = "Artist",
        date = "2026-07-20",
        quality = "1080p",
        format = "mp4",
        id = "abc123",
    )

    @Test
    fun `substitutes all variables`() {
        val result = FilenameTemplate.render("{channel} - {title} [{quality}]", values, 250)
        assertEquals("Artist - Song Title [1080p]", result)
    }

    @Test
    fun `removes characters android rejects`() {
        val result = FilenameTemplate.render("{title}", values.copy(title = "a/b:c*?\"<>|d"), 250)
        assertTrue(result.none { it in "\\/:*?\"<>|" })
    }

    @Test
    fun `validate rejects unknown variable`() {
        val validation = FilenameTemplate.validate("{title}-{bogus}")
        assertTrue(validation is FilenameTemplate.Validation.Invalid)
    }

    @Test
    fun `validate rejects unbalanced braces`() {
        assertTrue(FilenameTemplate.validate("{title") is FilenameTemplate.Validation.Invalid)
        assertTrue(FilenameTemplate.validate("title}") is FilenameTemplate.Validation.Invalid)
    }

    @Test
    fun `validate accepts a well formed template`() {
        assertEquals(FilenameTemplate.Validation.Valid, FilenameTemplate.validate("{title} {id}"))
    }

    @Test
    fun `render falls back to title for invalid template`() {
        val result = FilenameTemplate.render("{unknown}", values, 250)
        assertEquals("Song Title", result)
    }

    @Test
    fun `clamp preserves a trailing extension`() {
        val long = "x".repeat(300) + ".mp4"
        val result = FilenameTemplate.render(long, values, 20)
        assertTrue(result.endsWith(".mp4"))
        assertTrue(result.length <= 20)
    }

    @Test
    fun `render never returns blank`() {
        val result = FilenameTemplate.render("{title}", values.copy(title = "***"), 120)
        assertTrue(result.isNotBlank())
    }

    @Test
    fun `sanitize collapses repeated underscores and trims`() {
        assertEquals("a_b", FilenameTemplate.sanitize("__a///b__"))
    }

    @Test
    fun `example is not blank`() {
        assertTrue(FilenameTemplate.example("{title}-{channel}", 120).isNotBlank())
    }
}
